### PR TITLE
Undo Redo for Move page and Change next page

### PR DESCRIPTION
**Pull request by H3ll3m4**
[Trello](https://trello.com/c/rMBqTlnQ/2886-undo-redo-rails-ootb)

## Context
We want to provide the user the possibility to undo some actions. It is possible to get the previous version of metadata by calling the previous methods on the metadata api endpoint. This allow to get the the previous version of the form.

## Scope
Scope is limited to two user actions Move page and Change Next Page.

## Implementation
Move page is handled by Move controller and Change next page by Destinations Controller. Once the user choose to undo or redo an action, this action is passed as params to the Undo Controller. The Undo Controller handle the calls to the previous version from the Metadata API Client. It also store current undo/redo action information in the session. Then the Service Controller pass this information into a button for the view to display. 

## Tests
Adding unit tests, integration tests as request spec and acceptance tests.

### CODE REVIEW

File: app/controllers/api/move_controller.rb [hunk 1]
```diff
@@ -8,6 +8,15 @@ def targets
...
       @move.change
+
+      if @move.valid?
+        session[:undo] = {
+          action: 'undo',
+          undoable_action: 'move',
+          text: t('actions.undo_redo.undo_move')
+        }
+      end
+      # TODO: do we want to add here in this path toos
```
Comment (brenetic) [inline app/controllers/api/move_controller.rb:19]: do we need this?

File: app/controllers/api/destinations_controller.rb [hunk 2]
```diff
@@ -8,6 +8,11 @@ def new
 
     def create
       destination.change
+      session[:undo] = {
+        action: 'undo',
+        undoable_action: 'change_next_page',
```
Comment (brenetic) [inline app/controllers/api/destinations_controller.rb:13]: i read `undoable` as something that cannot be undone 😆

File: app/controllers/api/move_controller.rb (hunk 1 above)
Comment (brenetic) [inline app/controllers/api/move_controller.rb:19]: do we need this comment?

File: app/controllers/api/undo_controller.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+          text: t('actions.undo_redo.undo_change_next_page')
+        }
+      when 'move'
+        session[:undo] = {
+          action: 'undo',
+          undoable_action: 'move',
+          text: t('actions.undo_redo.undo_move')
+        }
+      end
+      previous
```
Comment (brenetic) [inline app/controllers/api/undo_controller.rb:18]: can we name this something a little more descriptive please

**Comment by brenetic** [comment]
Coolio, couple of comments. 

Seeing all the logic in the controller now I realise that that's a fair amount of logic. How would you feel about breaking it out into a class? For use in both the `undo` and `undo_redo` methods. If so that could mean a reduction in request specs and just have unit tests instead.

Also please do squash those commits down. They probably don't need to be that granular.

Other than that, great stuff! 👍

File: app/controllers/services_controller.rb [hunk 4]
```diff
@@ -17,6 +17,8 @@ def create
   end
 
   def edit
+    @undo_redo_button = session.delete(:undo)
```
Comment (brenetic) [inline app/controllers/services_controller.rb:20]:
```suggestion
    if ENV['undo_redo'] == 'enabled'
      @undo_redo_button = session.delete(:undo)
    end
```

assuming you don't want to launch the feature without the error messages?

Comment (H3ll3m4) [reply app/controllers/services_controller.rb:20]: Added feature flag

File: app/controllers/api/undo_controller.rb (hunk 3 above)
Comment (H3ll3m4) [reply app/controllers/api/undo_controller.rb]: Changed to call_previous_version

File: app/controllers/api/move_controller.rb (hunk 1 above)
Comment (H3ll3m4) [reply app/controllers/api/move_controller.rb]: comments removed

File: app/controllers/api/destinations_controller.rb (hunk 2 above)
Comment (H3ll3m4) [reply app/controllers/api/destinations_controller.rb:13]: will keep until we find a better name

File: app/controllers/api/move_controller.rb (hunk 1 above)
Comment (H3ll3m4) [reply app/controllers/api/move_controller.rb]: comments removed

**Review by brenetic — APPROVED** [review]
🚀

**Pull request merged**